man/man1/bdf2wfnt.1
.TH BDF2WFNT 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
bdf2wfnt \- convert BDF fonts to W font format
.SH SYNOPSIS
bdf2wfnt [\fIoptions\fP] [\fIBDF file\fP]
.SH DESCRIPTION
\fIbdf2wfnt\fP is an utility for converting X Window System BDF fonts to
W font format. \fIbdf2wfnt\fP reads a BDF font from file or standard input
and outputs a W font to standard output.
.SH OPTIONS
.TP
-m
Try to fit characters into as small space as possible by ignoring
the character descent and x-offset in the bounding box. Most likely
some characters won't quite align along the baseline, but you can correct
that with \fIfontedit\fP.  The resulting character height will then be
about the same as BDF font PIXEL_SIZE.
.TP
-v
Output information all the characters converted.
.PP
Any other option shows the program usage.
.SH BDF FONTS
You can acquire BDF fonts from your X server by running first a
\fIxfs\fP font server and then querying it with \fIfstobdf\fP.  You can
check out the names of available fonts with \fPxlsfonts\fP, their looks
with \fIxfontsel\fP and the whole charsets with \fIxfd\fP.
.SH EXAMPLES
First run the X font server:
.br
	xfs &
.PP
and then get one font like this:
.br
	fstobdf -server tcp/localhost:7100 -fn font
.PP
Just replace 'font' with a real X font name
like '-misc-fixed-medium-r-normal--7-70-75-75-c-50-iso8859-1'
or a wildcard abbreviation like '-*-fixed-*-r-*-*-7-*'.
.PP
If you'll pipe \fIfstobdf\fP output through \fIbdf2wfnt\fP and
\fIfontedit\fP (use '-' option for stdin), the font will be saved to
current directory with a suitable name for the W server.  As X font
family name may not be suitable for W, you could set a new one with
\fIfontedit\fP '-n <name>' option.  For example I prefer 'book' instead
of 'New Century Schoolbook', it's so much easier to write...
.SH HISTORY
\fIbdf2wfnt\fP is derived from the \fIbdf2c\fP program by Andrew
Scherpbier.
.SH SEE ALSO
.IR fontedit (1),
.IR fstobdf (1),
.IR xfd (1),
.IR xfs (1),
.IR xfontsel (1),
.IR xlsfonts (1)
.SH LAST CHANGED
6th Mar 1998 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


fonts/bdf2wfnt.c
/* 
 * bdf2wfnt.c (w) 1998 by Eero Tamminen
 * 
 * An X BDF font description to W font conversion tool.
 *
 * NOTES:
 * - Font size is according to it's bounding box unless '-m' option is used.
 * - BDF 'black' & 'bold' weights are mapped to F_BOLD and italic' &
 *   'oblique' styles are mapped to F_ITALIC.  No other styles are mapped.
 * - Font copyright and notice information are saved to font info.
 * - X font family name may exceed the space in W font header.  if it does,
 *   the name is truncated.  ATM bdf2wfnt preserves the case of name.
 * - Characters between codes 32-126 are identified according to their X
 *   name, rest of the chars are identified by their ENCODE line, if it
 *   follows immedietly the STARTCHAR line and has a legal value (1-255).
 *
 * Based on heavily modified and expanded bdf2c.c source Copyright (c) 1994
 * by Andrew Scherbier. See file 'COPYRIGHT'.
 */

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/in.h>
#include "../server/types.h"	/* font file header */
#include "Wlib.h"		/* font flags */


#define BUFFER_SIZE	256	/* for bdf input lines */


#define MAGIC	0x57464e31	/* W font file ID */
typedef struct
{
  FONTHDR header;
  char *info;			/* font info (author etc.) */
  uchar widths[256];		/* induvidual character widths */
  ulong *data;			/* long aligned character bitstream */
  ulong *map;			/* font bitmaps */
} MEMFONT;

static int verbose, minimal;	/* options */
static int ascent, descent;	/* additional info */


/* X character names */
#define START_INDEX	32
static char *charmap[] = {
	"space",
	"exclam",
	"quotedbl",
	"numbersign",
	"dollar",
	"percent",
	"ampersand",
	"quoteright",
	"parenleft",
	"parenright",
	"asterisk",
	"plus",
	"comma",
	"minus",
	"period",
	"slash",
	"zero",
	"one",
	"two",
	"three",
	"four",
	"five",
	"six",
	"seven",
	"eight",
	"nine",
	"colon",
	"semicolon",
	"less",
	"equal",
	"greater",
	"question",
	"at",
	"A",
	"B",
	"C",
	"D",
	"E",
	"F",
	"G",
	"H",
	"I",
	"J",
	"K",
	"L",
	"M",
	"N",
	"O",
	"P",
	"Q",
	"R",
	"S",
	"T",
	"U",
	"V",
	"W",
	"X",
	"Y",
	"Z",
	"bracketleft",
	"backslash",
	"bracketright",
	"asciicircum",
	"underscore",
	"quoteleft",
	"a",
	"b",
	"c",
	"d",
	"e",
	"f",
	"g",
	"h",
	"i",
	"j",
	"k",
	"l",
	"m",
	"n",
	"o",
	"p",
	"q",
	"r",
	"s",
	"t",
	"u",
	"v",
	"w",
	"x",
	"y",
	"z",
	"braceleft",
	"bar",
	"braceright",
	"asciitilde",
	NULL
};


/* 
 * read character from BDF file.  return 1 for success, 0 for failure.
 */
static int convert_char(MEMFONT *font, char *name, FILE *fp)
{
	static char buffer[BUFFER_SIZE];

	/* character width/height, bitmap width/height, offsets to baseline */
	int width, height, wd, ht, yoff, xoff;
	int idx, read_data, line, count;
	ulong value, *data;
	char *ptr;

	/*
	 * Lookup the character index
	 */
	for (idx = 0; charmap[idx]; idx++) {
		if (strcmp(name, charmap[idx]) == 0)
			break;
	}

	if (charmap[idx]) {
		idx += START_INDEX;
	} else {
		/* unrecogniced name, try to get encoding index */
		if (fgets(buffer, sizeof(buffer), fp) &&
		    strncmp(buffer, "ENCODING", 8) == 0) {
			sscanf(buffer+9, "%d", &idx);
			if (idx <= 0 || idx > 255) {
				return 0;		/* Not found */
			}
		}
	}

	height = font->header.height;
	data = font->map + idx * height;
	read_data = line = width = wd = ht = yoff = xoff = 0;

	if (verbose) {
		fprintf(stderr, "character: '%c' (%d)\n", idx, idx);
	}

	while (fgets(buffer, sizeof(buffer), fp)) {

		/*
		 * Read BITMAP data
		 */
		if (read_data) {
			if (strncmp(buffer, "ENDCHAR", 7) == 0) {
				return 1;
			}
			if (line++ >= height) {
				fprintf(stderr, "BBX <-> BITMAP inconsistency\n");
				return 1;
			}

			ptr = buffer;
			if (!isxdigit(*ptr)) {
				continue;
			}
			count = 0;
			value = 0;
			for (;;) {
				if (isdigit(*ptr)) {
					value |= (*ptr - '0') & 15;
				} else {
					value |= (*ptr - 'A' + 10) & 15;
				}
				count += 4;
				ptr++;
				if (!isxdigit(*ptr)) {
					break;
				}
				value <<= 4;
			}
			if ((count = count - wd) > 0) {
				value >>= count;
			}
			/* not an 'italic correction'? */
			if (xoff > 0) {
				value <<= xoff;
			}
			*data++ = value;
			continue;
		}

		/*
		 * Search for the BITMAP line
		 */
		if (strncmp(buffer, "BITMAP", 6) == 0) {
			if (verbose) {
				fprintf(stderr,
					"wd: %d, ht: %d, xoff: %d, yoff: %d, width: %d\n",
					wd, ht, xoff, yoff, width);
			}
			if (width > wd) {
				/* don't know whether these help anything... */
				if (xoff > 0) {
					if (minimal) {
						width -= xoff;
					}
					xoff = width - wd - xoff;
				} else {
					xoff = width - wd + xoff;
				}
			} else {
				width = wd;
			}
			if (width) {
				font->widths[idx] = width;
			} else {
				fprintf(stderr, "char has no width!\n");
				return 0;
			}
			line = height + descent - ht - yoff;

			if (line < 0 || line + ht > height) {
				fprintf(stderr, "ascent/descent error\n");
				if (line > 0) {
					line = height - ht;
				} else {
					line = 0;
				}
			}
			data += line;
			read_data = 1;

		} else if (strncmp(buffer, "BBX", 3) == 0) {
			sscanf(buffer+4, "%d %d %d %d", &wd, &ht, &xoff, &yoff);

		} else if (strncmp(buffer, "DWIDTH", 6) == 0) {
			sscanf(buffer+7, "%d", &width);
		}
	}

	fprintf(stderr, "abrupt bdf file end!\n");
	return 0;
}


static char *skip_space(char *token)
{
	while (*token && *token <= ' ') {
		token++;
	}
	return token;
}

static char *add_info(char *old, char *new)
{
	size_t len;
	char *info;

	if (!new) {
		return old;
	}
	if (!old) {
		return strdup(new);
	}
	free(old);
	len = strlen(old);
	info = malloc(len + strlen(new) + 2);
	if (!info) {
		fprintf(stderr, "allocating space for '%s' + '%s' failed\n",
			old, new);
		return NULL;
	}
	strcpy(info, old);
	info[len++] = ' ';
	strcpy(info+len, new);
	return info;
}

/*
 * return number of parsed characters. Types pointed by args may all
 * be modified.
 */
static int parse_bdf(MEMFONT *font, FILE *fp, char *buffer)
{
	FONTHDR *header = &(font->header);
	char *token;

	token = strtok(buffer, " \r\n");

	/* character info / bitmap */
	if (strcmp(token, "STARTCHAR") == 0) {
		if (!font->map) {
			long size = header->height;

			if (!size) {
				fprintf(stderr, "font has no height!\n");
				exit(-1);
			}
			/* with the height we can allocate bitmaps */
			font->map = calloc(256, size * sizeof(long));
		}
		token = strtok(NULL, " \r\n");
		return convert_char(font, token, fp);
	}

	/* font size */
	if (strcmp(token, "FONTBOUNDINGBOX") == 0) {
		if (font->map) {
			return 0;
		}
		token = strtok(NULL, "\r\n");
		sscanf (token, "%hd %hd %d %d",
			&(header->maxwidth), &(header->height),
			&ascent, &descent);

		if (minimal) {
			/* most characters should fit this quite OK too */
			header->height += descent;
		}
		header->baseline = header->height + descent;

	/* font family */
	} else if (strcmp(token, "FAMILY_NAME") == 0) {
		token = strtok(NULL, " \"\r\n");
		strncpy(header->family, token, sizeof(header->family));

	/* font styles */
	} else if (strcmp(token, "WEIGHT_NAME") == 0) {
		token = strtok(NULL, " \"\r\n");
		if (strcasecmp(token, "black") == 0 ||
		    strcasecmp(token, "bold") == 0) {
			header->styles |= F_BOLD;
		}
	} else if (strcmp(token, "SLANT") == 0) {
		token = strtok(NULL, " \"\r\n");
		/* IMHO X Italic and Oblique can both be mapped to W italic */
		if (*token == 'i' || *token == 'o' ||
		    *token == 'I' || *token == 'O') {
			header->styles |= F_ITALIC;
		}

	/* font info <- copyright / notices */
	} else if (strcmp(token, "COPYRIGHT") == 0 ||
	           strcmp(token, "NOTICE") == 0) {
		/* not sure if this works with all strtok()
		 * versions...
		 */
		token = skip_space(token + strlen(token) + 1);
		token = strtok(token, "\"\r\n");
		if (*token) {
			font->info = add_info(font->info, token);
		}
	}

	return 0;
}


/* 
 * W font saving utility functions
 */

/* set font maxwidth and flags according to character widths */
static void check_widths(MEMFONT *font)
{
	int idx, width, start = 32, end = 127, max = 0;

	for (idx = 0; idx < 32; idx++) {
		if (font->widths[idx]) {
			start = idx;
			break;
		}
	}
	for (idx = 255; idx > 127; idx--) {
		if (font->widths[idx]) {
			end = idx;
		}
	}

	font->header.flags &= ~(F_ASCII | F_LATIN1 | F_SYMBOL | F_PROP);

	if (start < 32) {
		/* symbols, native DOS, Mac, Atari... fonts */
		font->header.flags |= F_SYMBOL;
	} else {
		if (end > 127) {
			font->header.flags |= F_LATIN1;
		} else {
			font->header.flags |= F_ASCII;
		}
	}

	for(idx = start; idx <= end; idx++) {
		width = font->widths[idx];
		if(width > max) {
			if(max) {
				font->header.flags |= F_PROP;
			}
			max = width;
		}
	}
	font->header.maxwidth = max;
}

static ulong *map2stream(MEMFONT *font)
{
	register ulong *cptr, cbit, cdata, *dptr, dbit, ddata;
	register short cheight, lcwidth;
	short cwidth, height;
	ushort idx;
	long longs;

	if (font->data) {
		free(font->data);
	}
	longs = 0;
	height = font->header.height;
	for (idx = 0; idx < 256; idx++) {
		longs += (font->widths[idx] * height + 31) >> 5;
	}
	longs = longs * sizeof(long);
	font->data = malloc(longs);
	font->header.lenght = longs;

	if (!(font->map && font->data)) {
		fprintf(stderr, "font bitmap streaming failed!\n");
		return NULL;
	}

	memset(font->data, 0, longs);
	dptr = font->map;
	cptr = font->data;
	cbit = 0;

	for (idx = 0; idx < 256; idx++) {
		cwidth = font->widths[idx];
		if (!cwidth) {
			dptr += height;
			continue;
		}

		cdata = 0;
		cbit = 0x80000000;
		cheight = height;
		while (--cheight >= 0) {
			dbit = 1 << (cwidth-1);
			ddata = *dptr;
			lcwidth = cwidth;
			while (lcwidth-- > 0) {
				if (ddata & dbit) {
					cdata |= cbit;
				}
				dbit >>= 1;
				if (!(cbit >>= 1)) {
					cbit = 0x80000000;
					*cptr++ = cdata;
					cdata = 0;
				}
			}
			dptr++;
		}
		if (cbit != 0x80000000) {
			*cptr++ = cdata;
		}
	}
	free(font->map);
	font->map = NULL;
	return font->data;
}

/* Convert header variable byte orders between host<->big endian. */
static void convert_header(FONTHDR *header)
{
	/* font header / data size */
	header->magic  = ntohl(header->magic);
	header->lenght = ntohl(header->lenght);
	header->header = ntohs(header->header);

	/* font type information */
	header->height = ntohs(header->height);
	header->flags  = ntohs(header->flags);
	header->styles = ntohs(header->styles);

	/* font effect generation vars */
	header->thicken     = ntohs(header->thicken);
	header->skew        = ntohs(header->skew);
	header->effect_mask = ntohs(header->effect_mask);

	/* font character cell size */
	header->baseline = ntohs(header->baseline);
	header->maxwidth = ntohs(header->maxwidth);
}

static void save_font(MEMFONT *font)
{
	long idx, longs, info_len = 0;
	FONTHDR *header = &font->header;

	if (font->info && *(font->info)) {
		info_len = strlen(font->info) + 1;
	}

	/* set default values */
	header->magic = MAGIC;
	header->header = sizeof(FONTHDR) + info_len;
	header->effect_mask = ~header->styles & F_STYLEMASK;
	header->thicken = 1;
	header->skew = 0xAAAA;

	/* sets header->flags */
	check_widths(font);

	/* convert font data to bitstream, sets header->lenght */
	if (!map2stream(font)) {
		fprintf(stderr, "font bitstreaming failed!\n");
		return;
	}
	longs = font->header.lenght / sizeof(long);

	/* save header... */
	convert_header(&font->header);
	fwrite(&font->header, 1, sizeof(FONTHDR), stdout);
	if (info_len) {
		fwrite(font->info, 1, info_len, stdout);
	}
	fwrite(font->widths, 1, 256, stdout);

	/* ...and data */
	for (idx = 0; idx < longs; idx++) {
		font->data[idx] = htonl(font->data[idx]);
	}
	idx = fwrite(font->data, sizeof(long), longs, stdout);
	if (idx != longs) {
		fprintf(stderr, "font save failed!\n");
	}
}


/*
 * Get BDF font info from file and call above functions accordingly...
 */

static void usage(void)
{
	fprintf(stderr, "\
usage: bdf2wfnt [options] [BDF file]  >  <W font file>\n\
options:\n\
	-v  report conversions\n\
	-m  minimal size\n\
	-h  this help\n\
if there's no BDF file, stdin is read instead.\n");
}

int main(int argc, char **argv)
{
	static MEMFONT font;
	static char buffer[BUFFER_SIZE];
	int characters = 0, index = 0;
	FILE *fp;

	while(++index < argc && argv[index][0] == '-' && !argv[index][2]) {
		switch (argv[index][1]) {
			case 'm':
				minimal = 1;
				break;
			case 'v':
				verbose = 1;
				break;
			default:
				usage();
				return -1;
		}
	}

	if (index < argc) {
		if (!(fp = fopen(argv[index], "rb"))) {
			fprintf(stderr, "no bdf file!\n");
			usage();
			return -1;
		}
	} else {
		fp = stdin;
	}

	/* get attributes, character bitmaps etc. */
	while (fgets(buffer, sizeof(buffer), fp)) {
		characters += parse_bdf(&font, fp, buffer);
	}

	if (verbose) {
		fprintf(stderr,	"font: '%s'\n", argv[index]);
		fprintf(stderr,	"wd: %d, ht: %d, ascent: %d, descent: %d\n",
			font.header.maxwidth, font.header.height,
			ascent, descent);
		}

	if (characters) {
		save_font(&font);
	} else {
		fprintf(stderr, "no characters converted!\n");
	}
	return 0;
}

